runtime/near-vm-runner/src/runner.rs

```diff
@@ -105,7 +111,7 @@ pub trait VM {
         code: &[u8],
         code_hash: &CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Option<VMError>;
+    ) -> Option<Result<FunctionCallError, VMRunnerError>>;
```

Comment:
I think there's couple of problem here: 

* this should return CompilationError
* impls should use `?` to yeet VMRunnerError
* most critically, wasmtime should VMRunerError, not compilation error:

```diff
diff --git a/runtime/near-vm-errors/src/lib.rs b/runtime/near-vm-errors/src/lib.rs
index 95621e7c7..ab52b08c2 100644
--- a/runtime/near-vm-errors/src/lib.rs
+++ b/runtime/near-vm-errors/src/lib.rs
@@ -25,6 +25,9 @@ pub enum VMRunnerError {
     WasmUnknownError {
         debug_message: String,
     },
+    UnsupportedCompiler {
+        msg: String,
+    },
 }
 
 /// Permitted errors that cause a function call to fail gracefully.
@@ -151,7 +154,6 @@ pub enum CompilationError {
     CodeDoesNotExist { account_id: AccountId },
     PrepareError(PrepareError),
     WasmerCompileError { msg: String },
-    UnsupportedCompiler { msg: String },
 }
 
 #[derive(
@@ -394,9 +396,6 @@ impl fmt::Display for CompilationError {
             CompilationError::WasmerCompileError { msg } => {
                 write!(f, "Wasmer compilation error: {}", msg)
             }
-            CompilationError::UnsupportedCompiler { msg } => {
-                write!(f, "Unsupported compiler: {}", msg)
-            }
         }
     }
 }
@@ -419,6 +418,9 @@ impl fmt::Display for VMRunnerError {
             VMRunnerError::Nondeterministic(msg) => {
                 write!(f, "Nondeterministic error during contract execution: {}", msg)
             }
+            VMRunnerError::UnsupportedCompiler { msg } => {
+                write!(f, "Unsupported compiler {}", msg)
+            }
         }
     }
 }
diff --git a/runtime/near-vm-runner/src/runner.rs b/runtime/near-vm-runner/src/runner.rs
index 8776d95b7..c8858a4e5 100644
--- a/runtime/near-vm-runner/src/runner.rs
+++ b/runtime/near-vm-runner/src/runner.rs
@@ -5,7 +5,7 @@ use near_primitives::hash::CryptoHash;
 use near_primitives::runtime::fees::RuntimeFeesConfig;
 use near_primitives::types::CompiledContractCache;
 use near_primitives::version::ProtocolVersion;
-use near_vm_errors::{FunctionCallError, VMRunnerError};
+use near_vm_errors::{CompilationError, VMRunnerError};
 use near_vm_logic::types::PromiseResult;
 use near_vm_logic::{External, VMContext, VMOutcome};
 
@@ -117,7 +117,7 @@ pub trait VM {
         code: &[u8],
         code_hash: &CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Result<Option<FunctionCallError>, VMRunnerError>;
+    ) -> Result<Option<CompilationError>, VMRunnerError>;
 
     /// Verify the `code` contract can be compiled with this `VM`.
     ///
diff --git a/runtime/near-vm-runner/src/wasmer2_runner.rs b/runtime/near-vm-runner/src/wasmer2_runner.rs
index be0f5f731..8b7ceeba4 100644
--- a/runtime/near-vm-runner/src/wasmer2_runner.rs
+++ b/runtime/near-vm-runner/src/wasmer2_runner.rs
@@ -528,14 +528,14 @@ impl crate::runner::VM for Wasmer2VM {
         code: &[u8],
         code_hash: &near_primitives::hash::CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Result<Option<near_vm_errors::FunctionCallError>, VMRunnerError> {
-        let result = crate::cache::wasmer2_cache::compile_and_serialize_wasmer2(
+    ) -> Result<Option<near_vm_errors::CompilationError>, VMRunnerError> {
+        let outcome = crate::cache::wasmer2_cache::compile_and_serialize_wasmer2(
             code,
             code_hash,
             &self.config,
             cache,
-        );
-        let outcome = into_vm_result(result)?;
+        )
+        .map_err(VMRunnerError::CacheError)?;
         Ok(outcome.err())
     }
 
diff --git a/runtime/near-vm-runner/src/wasmer_runner.rs b/runtime/near-vm-runner/src/wasmer_runner.rs
index 84f40cf5a..d9ae80fee 100644
--- a/runtime/near-vm-runner/src/wasmer_runner.rs
+++ b/runtime/near-vm-runner/src/wasmer_runner.rs
@@ -327,14 +327,14 @@ impl crate::runner::VM for Wasmer0VM {
         code: &[u8],
         code_hash: &near_primitives::hash::CryptoHash,
         cache: &dyn CompiledContractCache,
-    ) -> Result<Option<FunctionCallError>, VMRunnerError> {
-        let result = crate::cache::wasmer0_cache::compile_and_serialize_wasmer(
+    ) -> Result<Option<CompilationError>, VMRunnerError> {
+        let outcome = crate::cache::wasmer0_cache::compile_and_serialize_wasmer(
             code,
             &self.config,
             code_hash,
             cache,
-        );
-        let outcome = into_vm_result(result)?;
+        )
+        .map_err(VMRunnerError::CacheError)?;
         Ok(outcome.err())
     }
 
diff --git a/runtime/near-vm-runner/src/wasmtime_runner.rs b/runtime/near-vm-runner/src/wasmtime_runner.rs
index 88f237ced..d52a45296 100644
--- a/runtime/near-vm-runner/src/wasmtime_runner.rs
+++ b/runtime/near-vm-runner/src/wasmtime_runner.rs
@@ -294,10 +294,10 @@ impl crate::runner::VM for WasmtimeVM {
         _code: &[u8],
         _code_hash: &CryptoHash,
         _cache: &dyn CompiledContractCache,
-    ) -> Result<Option<near_vm_errors::FunctionCallError>, VMRunnerError> {
-        Ok(Some(FunctionCallError::CompilationError(CompilationError::UnsupportedCompiler {
+    ) -> Result<Option<CompilationError>, VMRunnerError> {
+        Err(VMRunnerError::UnsupportedCompiler {
             msg: "Precompilation not supported in Wasmtime yet".to_string(),
-        })))
+        })
     }
 
     fn check_compile(&self, code: &[u8]) -> bool {
diff --git a/runtime/runtime/src/actions.rs b/runtime/runtime/src/actions.rs
index ed3629b1b..27c93c721 100644
--- a/runtime/runtime/src/actions.rs
+++ b/runtime/runtime/src/actions.rs
@@ -152,6 +152,9 @@ pub(crate) fn execute_function_call(
         VMRunnerError::WasmUnknownError { debug_message } => {
             panic!("Wasmer returned unknown message: {}", debug_message)
         }
+        VMRunnerError::UnsupportedCompiler { msg } => {
+            panic!("Wasmer returned unknown message: {}", msg)
+        }
     })
 }
 
```

Reply:
Agreed, this is how the errors should work here.

One problem, I noticed that `UnsupportedCompilerError` is part of the RPC schema:
https://github.com/near/nearcore/blob/26d1d7884a78f4e3759295559c0390f21f8aebac/chain/jsonrpc/res/rpc_errors_schema.json#L47-L56

Do we need to care about backward compatibility here? Or is this a dead error that never actually was stored anywhere?

Reply:
I *think* this is a dead error. I'd worry more about borsh than RPC schema, but I think we can be pretty sure that this is dead.

Reply:
I guess, we can leave `UnsupportedCompilerError` part of the change for a follow up as well?